Fix uniqueness validations on tables without a primary key

deivid-rodriguez opened the pull request:
If uniqueness validations are set up on a model whose corresponding table has been created with `id: false` or where `self.primary_key = nil` is specified, they will crash.

Commit:
Fix uniqueness validations on tables without a primary key

egilburg commented on activerecord/test/cases/validations/uniqueness_validation_test.rb:
```diff
@@ -35,6 +36,12 @@ class TopicWithUniqEvent < Topic
   validates :event, uniqueness: true
 end
 
+class MateyWithUniqValidation < Matey
```
All the specs here deal with uniq validation. This class specificially should be something like `MateyWithoutPrimaryKey`

sgrif replied:
Can we just define this class inline in the test?

sgrif commented on activerecord/test/cases/validations/uniqueness_validation_test.rb:
```diff
@@ -386,4 +393,13 @@ def test_validate_uniqueness_on_empty_relation
+    MateyWithUniqValidation.create(pirate_id: 1)
+    matey = MateyWithUniqValidation.create(pirate_id: 2)
+
+    matey.decrement(:pirate_id)
```
Do you think this test would be more clear if this and the previous line were written as:

```ruby
matey = MateyWithUniqValidation.new(pirate_id: 1)
```

deivid-rodriguez replied:
This is the only comment I don't agree with. Using `new` wouldn't persist the models so the uniqueness of the primary key validation would be skipped. Basically, the test would pass against the current code.

sgrif commented on activerecord/test/cases/validations/uniqueness_validation_test.rb:
```diff
@@ -5,6 +5,7 @@
 require 'models/warehouse_thing'
 require 'models/guid'
 require 'models/event'
+require 'models/matey'
```
How about we just set the table name on the subclass, rather than inheriting from this model, so we don't need to require more model files?

sgrif commented:
Looks good, just some style nit-picks. Please ping me once this has been updated.

deivid-rodriguez commented:
Hi @sgrif , thanks for the review, your idea of setting the table name in the model sounds great. Let me update the code.

deivid-rodriguez commented:
@sgrif done!

al2o3cr commented:
This change means that an already-saved record can't EVER be valid, since it will always find its own row.

For instance, changing the beginning of the test in this PR to:
```ruby
  def test_validate_uniqueness_on_model_without_primary_key
    first_record = ModelWithoutPrimaryKey.create!(pirate_id: 1)
    assert first_record.valid?, 'saved record should be valid if it succeeded'
```
makes it fail.

I'm unclear on whether this is actually a problem, as updating a record without a PK is impossible - it will break in `Relation#_update_record` [here](https://github.com/rails/rails/blob/dd53c0679d307378cfd7fadc67a5c180f2a61562/activerecord/lib/active_record/relation.rb#L83) for a very similar reason to this PR's issue. But it is, at minimum, an odd behavior.

deivid-rodriguez commented:
@al2o3cr You're totally right! I missed this...

I don't think It actually makes sense to add a uniqueness validation in a table without a primary key. The uniqueness of the whole set of columns (which is what I was trying to enforce here) is already guaranteed by default (a row in a relational database must be unique).

I think this can be closed actually...

al2o3cr commented:
@deivid-rodriguez Not sure what you mean. Without a unique index, the DB will happily accept new rows that match existing ones. I've had exactly that thing cause weird behavior with `has_and_belongs_to_many`.

I can't speak to your particular use case, but you may also want to consider something like the composite_primary_keys gem, if your table is really keyed by multiple columns taken together.

The "Rails way" would be to instead add an `id` column. YMMV as to whether that works with the constraints you face on your schema.

deivid-rodriguez commented:
@al2o3cr Thanks for the input and advice. 

I thought RDBMS's would prevent identical rows from being introduced in a table. I guess they keep an "internal" id instead and let everything in?

I run into this when I added a uniqueness constraint on the columns of a join table created with `id: false`. I think I'll just follow the Rails way and add an `id` column.

deivid-rodriguez commented:
To answer my own question and show my ignorance... SQL is not a proper relational language and SQL DBMS are not relational, so the "a row in a relational database must be unique" statement makes no sense here.

Back to the PR, I just rebased it and changed the test as @al2o3cr suggested to make it fail and it's still passing... What's going on?
